### PR TITLE
Enumerable Methods Project First Pull Request To Development Branch

### PR DESCRIPTION
**Pull request by KinezaEnaelle**
This project is about learning, understanding, and implementing some of the commonly used enumerable methods which include my_each, my_count, my_map, my_inject, my_select, my_any, my_none, my_all, and the use of block and proc.

### CODE REVIEW

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi 
Such a remarkable project you made :+1: ! Awesome README :100: !
We, however, do have a few changes to carry out before we proceed (see review comment).

##

 For this project, Kindly fix the failed test cases
- [ ] Rename `main.rb` to `enumerable.rb`
- [ ] Fix the linter issues
- [ ] Make the rest of the methods pass all the test cases provided below:

```
puts
puts "=========Test Cases========="
puts
puts "my_each"
print [1, 2, 3].my_each { |elem| print (elem + 1).to_s + " "} # => 2 3 4
puts

#2. my_each_with_index
puts
puts "my_each_with_index"
print [1, 2, 3].my_each_with_index { |elem, idx| puts "#{elem} : #{idx}" } # => 1 : 0, 2 : 1, 3 : 2
puts


# 4. my_all? (example test cases)
puts "my_all?"
test cases required by tse reviewer
p [1,2,3,4,5].my_all? # => true
p ['dog', 'door', 'rod', 'blade'].my_all?(/d/) # => true
p [1, 1, 1].my_all?(1) # => true
puts

#5. my_any? (example test cases)
puts "my_any?"
test cases required by tse reviewer
p [1, nil, false].my_any?(1) # => true
p ['dog', 'door', 'rod', 'blade'].my_any?(/z/) # => false
p [1, 2 ,3].my_any?(1) # => true
p %w[jes,umair,jesagain,hello].my_any?('jes') # => false
puts

#6. my_none? (example test cases)
puts "my_none?"
test cases required by tse reviewer
p [nil, false, nil, false].my_none? # => true
p [1, 2 ,3].my_none? # => false
p [1,2,3,4,5].my_none?(2) # => false
p [1, 2, 3].my_none?(4) # => true
puts

#7. my_count (example test cases)
puts "my_count"
p [1,2,3].my_count # => 3
p (1...5).my_count # => 4
puts

#8. my_map
puts "my_map"
p [1,2,3].my_map { |n| 2 * n } # => [2,4,6]
p ["Hey", "Jude"].my_map { |word| word + "?" } # => ["Hey?", "Jude?"]
p [false, true].my_map { |bool| !bool } # => [true, false]
puts

#9. my_inject
puts "my_inject"
p [1,2,3,4].my_inject { |accum, elem| accum + elem} # => 10
p [5, 1, 2].my_inject("+") # => 8
p (5..10).my_inject(2, :*) # should return 302400
p (5..10).my_inject(4) {|prod, n| prod * n} # should return 604800
```

##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    n = length
+    arr = *self
+    n.times do |i|
+      arr[i] = yield(arr[i])
+    end
+    self
+  end
```
Comment (peterrobert) [inline enumerable.rb]: Good work implementing some of the my_each functionalities. However, there are some changes to be made :arrow_down: 

- [x]  when a block is given when self is a range it should return the range itself after calling the given block once for each element in self

for example: if you have a range.my_each(&block) it should equal range.my_each(&block)

kindly reference here for the behaviour of the each method [link](https://apidock.com/ruby/Range/each)  `(10..15).each {|n| print n, ' ' }
# prints: 10 11 12 13 14 15`

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+  def my_each_with_index
+    return to_enum(:my_each) unless block_given?
+
+    n = length
+    n.times do |index|
+      yield(self[index], index)
+    end
+    self
+  end
```
Comment (peterrobert) [inline enumerable.rb]: - [x] when a block is given when self is a range it should return the range itself after calling the given block once for each element in self

EXAMPLE: range.my_each_with_index(&block)  `evens = (1..10).each_with_object([]) { |i, a| a << i*2 }
#=> [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]` use this [link](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index)

- [x] When you are given a Hash it should call the block once for each element in the hash. 

Example: `hash = Hash.new
%w(cat dog wombat).each_with_index { |item, index|
  hash[item] = index
}
hash   #=> {"cat"=>0, "dog"=>1, "wombat"=>2}`

- [ ]

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+  def my_select
+    arr = []
+    my_each do |elem|
+      arr << elem if yield(elem) == true
+    end
+    arr
+  end
```
Comment (peterrobert) [inline enumerable.rb]: - [x] First, it should return an enumerator if no block is given. [link](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select)

- [x] If you are given an array and the condition given returns true... Then it should return all the numbers that satisfy that condition.

EXAMPLE: `
[1,2,3,4,5].select { |num|  num.even?  }   #=> [2, 4]`

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      end
+    elsif !pattern.nil?
+      my_each do |elem|
+        unless (pattern == elem) || (pattern.is_a?(Class) && elem.is_a?(pattern)) || (pattern.is_a?(Regexp) && pattern.match?(elem.to_s))
+          return false
+        end
+      end
+    end
+    true
+  end
```
Comment (peterrobert) [inline enumerable.rb]: - [x] If the block given returns false or nil. Then it should return false

example: `[nil, true, 99].all?                              #=> false`

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED :recycle:  :stop_button: 

HELLO THERE :wave:  NICE PROGRESS WITH THE PROJECT :muscle:  YOU ARE MAKING IMPROVEMENTS AND THAT IS PERFECT :+1: 

HOWEVER, THERE ARE A FEW MORE CHANGES TO BE MADE KINDLY CHECK BELOW :arrow_down: 

- [ ] In your my any method kindly check how the behaviour should be when a range is given

- [ ] In your my none method kindly check how the behaviour should be when a range is given

### if you have any question kindly reach me on slack @peterrobert

### HAPPY CODING :computer:

File: enumerable.rb (hunk 2 above)
Comment (henatan99) [reply enumerable.rb]: Dear Peter, we appreciate the very helpful review you gave us and we addressed them. But with all due respect, the .each_with_object method is not a requirement for this project.

Comment (peterrobert) [reply enumerable.rb]: Hello @henatan99 sorry for the confusion. It is actually an error on my side. I actually meant each_with_index. Sorry again for the confusion and thank you for the polite correction.

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @henatan99 and @KinezaEnaelle 

Well done, advancing this far 👏 you still have some failing tests, please check them out:

### Methods:
  **Tip:** You can see the complete [description of all the methods here](https://ruby-doc.org/core-2.6.5/Enumerable.html), please check it out as a guide if you feel lost

#### #my_each_with_index
  - [x] Please check and correct the difference between original and your methods:
  ```
    p (1..10).each_with_index { |i, a| a << i*2 } #=> [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]
    p (1..10).my_each_with_index { |i, a| a << i*2 } #=> undefined method `[]' for 1..10:Range (NoMethodError)
  ```
  Notice that your code doesn't know how to iterate through the range **Hint:** You are trying to retrieve an index from a range, and that is not possible, check line 23

#### #my_any?
 - [ ] Please check and correct the difference between original and your methods:
  ```
    p [nil, false, true, []].any? #=> true
    p [nil, false, true, []].my_any? #=> false
  ```
  Notice that you are missing the case where there is no block and no pattern

#### #my_count
  - [x] `(1..3).my_count{|x| x >= 2}` should return 2 because there are 2 elements that have that condition in the block. Your method returns 5. **Hint** Your count is starting at count_item = size, so it is doing 3+2=5

#### #my_map
 - [x] Please check and correct the difference between original and your methods:
  ```
    p (2..5).map{ |x| x + 1 } #=> [3, 4, 5, 6]
    p (2..5).my_map{ |x| x + 1 } #=> undefined method `[]' for 2..5:Range (NoMethodError)
  ```
  Notice that your method should accept ranges **Hint:** this line needs to be fixed `arr[i] = yield(self[i])`

#### #my_inject
  - [x] If there is no block or argument, your method should return a local jump error ([please check how to raise errors here](https://stackoverflow.com/questions/4325507/how-to-force-a-localjumperror-on-return-in-ruby))

That would be it, good luck 🌟 

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Comment by kelvin8773** [comment]
# 👋 Change Required 😭  💅 

Hey, team,

Nice job to make it this far, let's keep it up! 💯 😄 

However, there is still 1 glitch needs to fix, so let's get into it. 🤝 

* **#my_any** - As defined in [Here](https://apidock.com/ruby/Enumerable/any%3F) or [Here](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-any-3F).
> Ruby adds an implicit block of { |obj| obj } that will cause #any? to return true if at least one of the collection members is not `false` or `nil`.

-  Which it means when `[nil, 99, false].any?` should return `true`.
-  While `[nil, false, nil].any` should return `false`.

But this is not the case for your `my_any` method yet, so please find a way to fix it, thanks!

That will be it!  Good luck and happy coding! 😄 👨‍💻 👩‍💻 

Best,

Kelvin

### Please Note:
Feel free to contact me at slack by @Kelvin or email [kelvinliang@yandex.com](kelvinliang@yandex.com) if you have any questions on this request, thanks!

**Review by kitpao — APPROVED** [review]
# 👋 Hello @KinezaEnaelle and @henatan99 

Great job making all your methods work as expected 🌟 

### 👏 Your project has been approved 👏
### 🚀 You may merge to development and submit the project completition form 🚀

___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Pull request merged**